test/Concurrency/Runtime/async_sequence.swift

```diff
@@ -0,0 +1,604 @@
+// RUN: %target-run-simple-swift(-Xfrontend -enable-experimental-concurrency)
+
+// REQUIRES: executable_test
+// REQUIRES: concurrency
+
+import StdlibUnittest
+
+// Utility functions for closure based operators to force them into throwing 
+// and async and throwing async contexts.
+
+func throwing<T>(_ value: T) throws -> T {
+  return value
+}
+
+func asynchronous<T>(_ value: T) async -> T {
+  return value
+}
+
+func asynchronousThrowing<T>(_ value: T) async throws -> T {
+  return value
+}
+
+struct Failure: Error { }
+
+func failable<T, E: Error>(
+  _ results: [Result<T, E>]
+) -> AsyncFailableMapSequence<AsyncLazySequence<[Result<T, E>]>, T> {
+  return results.async.map { try $0.get() }
+}
+
+
+extension Sequence {
+  @inlinable
+  public var async: AsyncLazySequence<Self> {
+    get {
+      return AsyncLazySequence(self)
+    }
+  }
+}
+
+@frozen
+public struct AsyncLazySequence<S: Sequence>: AsyncSequence {
+  public typealias Element = S.Element
+  public typealias AsyncIterator = Iterator
+
+  @frozen
+  public struct Iterator: AsyncIteratorProtocol {
+    @usableFromInline
+    var iterator: S.Iterator
+
+    @usableFromInline
+    init(_ iterator: S.Iterator) {
+      self.iterator = iterator
+    }
+
+    @inlinable
+    public mutating func next() async -> S.Element? {
+      return iterator.next()
+    }
+  }
+
+  public let sequence: S
+
+  @inlinable
+  public init(_ sequence: S) {
+    self.sequence = sequence
+  }
+
+  @inlinable
+  public __consuming func makeAsyncIterator() -> Iterator {
+    return Iterator(sequence.makeIterator())
+  }
+}
+
+extension AsyncSequence {
+  @inlinable
+  public func collect() async rethrows -> [Element] {
+    var items = [Element]()
+    var it = makeAsyncIterator()
+    while let e = try await it.next() {
+      items.append(e)
+    }
+    return items
+  }
+}
+
+var AsyncLazySequenceTests = TestSuite("AsyncLazySequence")
+
+AsyncLazySequenceTests.test("iteration past first nil") {
```

Comment:
100% agree! wasn't sure if scope creep of that form was something I should add to this commit or not (I am inclined to agree with ya after having written that...)

Reply:
I made a local variant of that since I wasn't sure if the TestSuite would ever grow proper async support or not (leaving that for someone else to do) But I refactored them all to funnel to an overload.